### ISSUE TITLE
spawned processes with lots of output get stuck

### ISSUE THREAD

**Issue by justin-stephenson** [cockpit-project/cockpit#12580]
Cockpit version: 199
OS: Fedora 29, Fedora 30
Page: Session Recording

cockpit-session-recording uses the journal.journalctl method to retrieve a list of journal entries, process them, then present a list of Recordings in the UI. Intermittently(about 40% of the time) when refreshing the main /session-recording page, only a subset of the total entries are returned to the .stream(data) handler. When this happens the .done() handler never is called. In our code we have:

        let matches = ["_EXE=/usr/bin/tlog-rec-session",
            "_EXE=/usr/bin/tlog-rec", "+",
            "_COMM=tlog-rec", "+",
            "_COMM=tlog-rec-session"];

        let options = {follow: false, count: "all", merge: true};

        this.journalctl = Journal.journalctl(matches, options)
                .fail(this.journalctlError)
                .stream(this.journalctlIngest)
                .done(() => {
		    console.log('Done!');
                });
    }

On my test system I have recordings(which consists of numerous back-to-back journal entries) containing a large amount of flooded output. I believe that cockpit.spawn(), when called by journalctl, is not processing all of the journalctl command JSON output. This leads to the lack of complete entries returned through journalctl, though I could not say why it happens only intermittently.

I also tried setting different cockpit.spawn option `batch` and `latency` values but the problem persists.

Steps to reproduce

1. dnf install tlog
2. tlog-rec -w journal cat /usr/share/dict/linux.words
3. In cockpit module code:

        let cmd = ["journalctl", "-q", "--lines=all", "_COMM=tlog-rec", "--output=json"];
        cockpit.spawn(cmd)
                .stream((data) => {
                    console.log(data);   /* If data contains Zyzzogeton, all entries were returned */
                    if (data.includes("Zyzzogeton")) {
                        console.log('Found string');
                    }
                })
                .done(() => {
                    console.log('Done!');
                });

4. Intermittently, not all of the output will be provided in .stream(data) and the .done() handler is never called

I can provide more information if needed, or feel free to contact me on IRC 'justintime'

Downstream report: https://bugzilla.redhat.com/show_bug.cgi?id=1751783

**Comment by justin-stephenson** [comment]
I believe the hanging output leading to missed entries can be reproduced in the Cockpit Terminal page with:
~~~
$ find /
~~~

**Comment by martinpitt** [comment]
Debugging notes:

 * The "find in terminal" command works really well to reproduce this. This can be done in a less I/O intense way with just running `seq 100000`. This works with `-s ' '` as well. The seq process then hangs on write(), i. e. its output buffer is full and does not get flushed. It works equally well in the headless chrome test. ([check-terminal test](https://github.com/martinpitt/cockpit/blob/spawn-flooding/test/verify/check-terminal#L98))

 * This does not seem to reproduce at all in the Firefox developer console. The dictionary/tlog test, find, and seq all work perfectly until the end.

 * Unfortunately this does not reproduce in a [unit test](https://github.com/martinpitt/cockpit/blob/spawn-flooding/src/base1/test-spawn-proc.js#L298), not even with one or two more magnitudes (which shovels 78 MB/888 MB of data through the channel). This also runs in the same chromium headless, but without routing this through cockpit-ws, just through test-server. I also tested this interactively in firefox, also succeeds.

 * I added a [spawn speed test](https://github.com/martinpitt/cockpit/blob/spawn-flooding/pkg/playground/speed.js#L198) to our playground. This works well with `seq`, but hangs with `find /`. It shows that `process.close()` still works and properly kills the process, so the channel isn't completely busted. It also demonstrates that this is sensitive to timing and/or I/O. On the terminal page, the rendering incurs significant slowdown, whereas the unit test just discards the output. But the playground works with (1) discarding, (2) console.log()'ing it, and (3) rendering it into a `<pre>`.

**Comment by martinpitt** [comment]
This is related to pressure handling. While reading data from `find`, normally this happens:
```
find: reading input from pid 22845 cond  1
stdio: queued 4096 byte payload
stdio: queued 53 byte payload
1:1: sending ping with sequence: 6310628
stdio: reading input from pid 0 cond  1
stdio: received a 54 byte payload
1:1: received pong with sequence: 6245092
stdio: wrote 4096 bytes
stdio: wrote 4 bytes
stdio: wrote 53 bytes
stdio: output queue empty
```

But at some point:
```
find: reading input from pid 22845 cond  1
stdio: queued 4096 byte payload
1:1: sent too much data without acknowledgement, emitting back pressure until 6318820
find: applying back pressure in pipe
stdio: wrote 9 bytes
stdio: wrote 4096 bytes
stdio: output queue empty
```

This is the last message from `find`. Afterwards there are still three pings, last one:
```
stdio: reading input from pid 0 cond  1
stdio: received a 54 byte payload
1:1: received pong with sequence: 6310628
stdio: want more data
```
After that, no more pings are received, thus the backpressure is never lifted.

Disabling the cockpit_flow_emit_pressure in cockpitpipe.c does not help. Disabling it in cockpitchannel.c does help (but is of course not a fix, just confirmation that it's *that* flow control)

**Comment by martinpitt** [comment]
@stefwalter : Would you be able to have a quick discussion with me how to fix the pressure protocol to avoid this deadlock?

**Comment by martinpitt** [comment]
I sent a proposed fix to #14079, which I'd like to discuss with @stefwalter , either in the PR or in a meeting.
